DDD-9.md

```diff
@@ -0,0 +1,111 @@
+
+```json
+{
+  "resume_scn": "123456",
+  "txId": "123456", // current transaction being processed
+  "txSeq": 25, // last event in current transaction received
+  "committed_transactions": [{
+    "id": "a748391bc3123",
+    "thread": 1,
+    "scn": "123478"
+  }]
+}
+```
+
+### LogMiner query
+
+The LogMiner query must be adjusted to include a new predicate condition to exclude transactions in `committed_transactions`.
+The query change would include something like the following:
+
+```sql
+AND UPPER(RAWTOHEX(XID)) NOT IN ( <committed-transaction.id(s) in upper-case> )
+```
+
+The construction of the in-clause should use the same logic for the `table.include.list` to wrap the predicate in a disjunction is the list of values exceeds 1000 values, since Oracle does not permit creating in-clauses with more than 1000 elements.
+
+In addition, the `RESUME_SCN` column should be included, if it isn't already.
+
+### Processor changes
+
+The processor logic needs to be changed to move away from using "caches", and instead will simply parse the `LogMinerEventRow` details, and dispatch.
+During the dispatch, the processor will likely need to be merged with the `TransactionCommitConsumer` to handle merging of CLOB, BLOB, XML, and Extended String operations on-the-fly.
+
+In addition, the returned `startScn` will no longer be based on the last read SCN, but instead will be based on the last non-null `getResumeScn()` value.
+As a value is present in the `getResumeScn()` event field, the offset's `resume_scn` should be updated.
+
+The `ResultSet` itself will always contain changes to transactions we've not yet seen and processed or a transaction that was mid-dispatch when the connector was stopped.
+If the transaction id matches the offset's `txId`, then the processor will ignore any event with a sequence less-than or equal-to the offset's `txSeq`, to avoid redispatching events already sent to Kafka.
+When a transaction commits, it is added to the offset's `committed_transactions` and the `txId` and `txSeq` are cleared.
+
+### Schema changes
+
+TBD
+
+## Concerns / Gaps
+
+* LogMiner will reorder all changes in chronological commit order, meaning SCN values will increment/decrement from row-to-row, while the commits will always be in chronological order, always equal-to or greater.
+* DDL changes need to be reviewed to make sure they do not introduce any unexpected/unintended consequences of dispatch/processing order.
+* Large transactions will be re-read across mining steps; will this create any disk IO performance concerns on the database?
+* Will this be a problem for any of the three current mining strategies using committed data?
+* What about partial rollbacks, do we still need to handle these?
+* Do we implement this as a new adapter, i.e. `logminer_slim`, or change `logminer` proper?
+* Is the change in offsets being non-backward compatible a blocker or concern?
```

Comment:
There should be a migration path from old offsets to the new adapter when old adapter is kept as default.
W'll need to discuss the need for new -> back migration path only when the new adapter will become new default.